source/h2_decoder.c

```diff
@@ -38,22 +39,60 @@ static const uint32_t s_31_bit_mask = UINT32_MAX >> 1;
+            if ((decoder)->vtable->fn((decoder)->userdata)) {                                                          \
+                DECODER_LOGF(ERROR, decoder, "Error from callback " #fn ", %s", aws_error_name(aws_last_error()));     \
+                return AWS_OP_ERR;                                                                                     \
+            }                                                                                                          \
         }                                                                                                              \
     } while (false)
 #define DECODER_CALL_VTABLE_ARGS(decoder, fn, ...)                                                                     \
     do {                                                                                                               \
         if ((decoder)->vtable->fn) {                                                                                   \
-            DECODER_LOG(TRACE, decoder, "Calling user callback " #fn)                                                  \
-            (decoder)->vtable->fn(__VA_ARGS__, (decoder)->userdata);                                                   \
+            DECODER_LOG(TRACE, decoder, "Invoking callback " #fn)                                                      \
+            if ((decoder)->vtable->fn(__VA_ARGS__, (decoder)->userdata)) {                                             \
+                DECODER_LOGF(ERROR, decoder, "Error from callback " #fn ", %s", aws_error_name(aws_last_error()));     \
+                return AWS_OP_ERR;                                                                                     \
+            }                                                                                                          \
         }                                                                                                              \
     } while (false)
 #define DECODER_CALL_VTABLE_STREAM(decoder, fn)                                                                        \
     DECODER_CALL_VTABLE_ARGS(decoder, fn, (decoder)->frame_in_progress.stream_id)
 #define DECODER_CALL_VTABLE_STREAM_ARGS(decoder, fn, ...)                                                              \
     DECODER_CALL_VTABLE_ARGS(decoder, fn, (decoder)->frame_in_progress.stream_id, __VA_ARGS__)
 
+/* for storing things in array without worrying about the specific values of the other AWS_HTTP_HEADER_XYZ enums */
+enum pseudoheader_name {
+    PSEUDOHEADER_UNKNOWN = -1, /* Unrecognized value */
+
+    /* Request pseudo-headers */
+    PSEUDOHEADER_METHOD,
+    PSEUDOHEADER_SCHEME,
+    PSEUDOHEADER_AUTHORITY,
+    PSEUDOHEADER_PATH,
+    /* Response pseudo-headers */
+    PSEUDOHEADER_STATUS,
+
+    PSEUDOHEADER_COUNT, /* Number of enums */
+};
+
+static enum pseudoheader_name s_header_to_pseudoheader_name(enum aws_http_header_name name) {
+    switch (name) {
```

Comment:
nit take it or leave, you could do this as an array and avoid the branch misses here and the code would be just as readable.

Reply:
Compilers are smart:
https://godbolt.org/z/6FpXN4
vs
https://godbolt.org/z/TFCk9A

I can't actually read assembly, but it looks like there are less branches in the switch version.